cheatsheets/DotNet_Security_Cheat_Sheet.md

```diff
@@ -493,13 +741,197 @@ After .NET Core 2.0 it is possible to automatically generate and verify the anti
+
+//Check below is too late! Execution may have already occurred.
+if (suspectObject is SomeDangerousObjectType) 
+{
+    //generate warnings and dispose of suspectObject
+}
+```    
+
+For `BinaryFormatter` and `JSON.Net` it is possible to create a safer form of white list control useing a custom `SerializationBinder`.
+
+Try to keep up-to-date on known .Net insecure deserialization gadgets and pay special attention where such types can be created by your deserialization processes. **A deserializer can only instantiate types that it knows about**. 
+
+Try to keep any code that might create potential gagdets separate from any code that has internet connectivity. As an example `System.Windows.Data.ObjectDataProvider` used in WPF applications is a known gadget that allows arbitrary method invocation. It would be risky to have this a reference to this assembly in a REST service project that deserializes untrusted data.
+
+### Known .NET RCE Gadgets
+
+- `System.Configuration.Install.AssemblyInstaller`
+- `System.Activities.Presentation.WorkflowDesigner`
+- `System.Windows.ResourceDictionary`
+- `System.Windows.Data.ObjectDataProvider`
+- `System.Windows.Forms.BindingSource`
+- `Microsoft.Exchange.Management.SystemManager.WinForms.ExchangeSettingsProvider`
+- `System.Data.DataViewManager, System.Xml.XmlDocument/XmlDataDocument`
+- `System.Management.Automation.PSObject`
+
+More information can be found here: [Deserialization Cheat Sheet](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Deserialization_Cheat_Sheet.md)
+
+## A9 Using Components with Known Vulnerabilities
 
 DO: Keep the .Net framework updated with the latest patches
 
 DO: Keep your [NuGet](https://docs.microsoft.com/en-us/nuget/) packages up to date, many will contain their own vulnerabilities.
 
-DO: Run the [OWASP Dependency Checker](https://www.owasp.org/index.php/OWASP_Dependency_Check) against your application as part of your build process and act on any high level vulnerabilities. 
+DO: Run the [OWASP Dependency Checker](https://www.owasp.org/index.php/OWASP_Dependency_Check) against your application as part of your build process and act on any high level vulnerabilities.
+
+## A10 Insufficient Logging & Monitoring
+
+DO: Ensure all login, access control failures and server-side input validation failures can be logged with sufficient user context to identify suspicious or malicious accounts, and are held for sufficient time to allow delayed forensic analysis.
+
+DO: Ensure logs are generated in a format that can be easily consumed by a centralized log management solution.
+
+DO: Ensure high-value transactions have an audit trail with integrity controls to prevent tampering or deletion, such as append-only database tables.
+
+DO: Establish effective monitoring and alerting so suspicious activities are detected and responded to in a timely fashion.
+
+DO: Establish or adopt an incident response and recovery plan, such as NIST 800-61 rev 2.
+
+### Logging
```

Comment:
This is copied from https://microsoft.github.io/code-with-engineering-playbook/Engineering/DevOpsLoggingDetailsCSharp.html. Please summarise it somehow, change example to something that you mentioned before (e.g. logging server-side input validation failures) and for logging levels add examples more security focused.

Reply:
When writing this, I completely missed the point. The logging I talk about here is application errors etc. Not security errors- such as SQL Validation, Login Attempts, or Cookie Tampering etc.... How much detail do you want to see here? would you want to see something like this? https://www.carlrippon.com/adding-useful-information-to-asp-net-core-web-api-serilog-logs/. Please advise @mackowski

Reply:
10 lines up you wrote "Ensure all login, access control failures and server-side input validation failures can be logged with sufficient user context to identify suspicious or malicious accounts, and are held for sufficient time to allow delayed forensic analysis" so try to make it consistent :) 
And application errors are also security related.

Reply:
I just think that you should summarise these articles and link to them for more information and if you want to put an example make it security related one.
CS should bring knowledge on specific application security topics. Developers who will read this document want to know how to implement [this](https://www.owasp.org/index.php/Top_10-2017_A10-Insufficient_Logging%26Monitoring) not how logging works in .NET in general. 
You can also check https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Logging_Cheat_Sheet.md and https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Error_Handling_Cheat_Sheet.md

For example try to summarise what should be logged and reference to other resources for details and than you can create related example and also reference to other resources for details about it.

Reply:
cool.. I shall try